trying to clean up

ckandola opened the pull request:
I got rid of the "magic number", and I tried to clean up the main for loop by using splice() to remove elements. But without removeDuplicates(), the output now includes duplicate city names. Other than that, I'm not sure how to make compareObjects() less expensive. I was using unicode so I could make 'len' and 'sum' properties that could be compared between objects to see if the same letters were used and if their unicode sums match.

laurenclarke commented:
@ckandola ok, we are going to camp here for a bit. This needs work. There's lots to learn here. 

I'd like you to re-think the sequence of work being done here. Are the places where you can improve things by doing them in different order?

Also, and probably more importantly, try to think of a way to establish equivalence between strings with respect to anagrams. 

Two hints:
1. You were onto something here... `cities[i].name.toUpperCase()` why did you do this? what were you trying to accomplish? what problem were you solving? do you have similar problems that could be solved in similar ways? 
2. You are writing javascript but thinking in java. You are getting tangled in nouns (objects and properties). 

Finally, please make your changes and post commits here so we can keep a running set of changes. My memory is shot, and I don't want to have to rummage in closed pull requests to find my previous comments ;).

ckandola commented:
Sorry, that last pull request was so badly done that I wanted to retry it. I'll stay on this one now.

I definitely agree that I've been thinking in a Java way. If I don't use objects, then should I just take each name and compare their letters to each other? Would it be Java-ish to make the strings into arrays and take off letters when there's a match?

laurenclarke commented:
@ckandola Ok, I think it's true that you can stay with strings (vs turning them into objects. I also think it's true that you don't have to do step-wise, letter-by-letter comparison (at least, that's not code you have to write).  The hint is: you don't have to compare the actual strings you are given, in fact, you are not doing that already. You've converted to upper-case representation since  you've decided case doesn't matter (and I agree). You've also stripped spaces. So, in the end you are not comparing the original strings, but cleaner _representations_ of them. What other pre-processing could you do to make anagram equivalence easy to find?

Sure, remove dupes (or suppress those matches).

ckandola commented:
Can I sort the array I made, or is that CPU-intensive?

I put the issue of removing all duplicates to the side for now.

laurenclarke commented:
Ok, yes. I think you are getting warmer. The idea is to mold the data into a shape that allows you to apply native (to javascript) array and string manipulation tools.
 
>Can I sort the array I made, or is that CPU-intensive?
Try it and see. If you like, you can get some insights from the tools in chrome: https://developers.google.com/web/tools/chrome-devtools/evaluate-performance/

>I put the issue of removing all duplicates to the side for now.
Ok. That's fine. 

I think it would be helpful for you to write out a plain-english pseudo-code description of your method. (You can do that here or in a gdoc). Let's look at that together.

ckandola commented:
@laurenclarke Okay, it feels pretty straight-forward now with just sort(), but I typed some pseudo-code of the comparing function here https://drive.google.com/open?id=1t2DgRyum7YG4bCY_2o_hAnLnAbYNNkAYhrYR0VEPYPU.

ckandola commented:
@laurenclarke  Okay, this is much faster with @prust 's suggestions. I also updated the [output file.](https://github.com/ckandola/javascript-problems/blob/second-attempt/anagrams/anagrams_output.txt)

laurenclarke commented on anagrams/anagram-cities.js:
```diff
@@ -7,111 +7,43 @@
+// for each city in the list, add an array with hashed version and index to the array
 for(var i = 0; i < cities_len; i++){
-  makeCharObj(cities[i].name.toUpperCase());
+  hashedNames[i] = [cities[i].name.toUpperCase().replace(/[\s\(\)\-]/g, '').split('').sort().join(''), i];
```
Yes!

FWIW, I think this would be a bit clearer if you spit it up into separate lines, but the action is correct.

laurenclarke commented:
@ckandola this is much better. Ok, here are some extensions to try: 
1. Can you modify so that spaces alone are not sufficient to "count" 
` [ 'Rocksprings', 'Rock Springs' ]`
2. Can you modify so that re-ordering of the same words are not considered
` [ 'Lake Forest', 'Forest Lake', 'Forest Lake', 'Forest Lake' ],`
3. Remove the dupes, or maybe better, count the repeated entries, so
``` 
[ 'Selma',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Salem',
    'Lasem',
    'Salem',
    'Elmas' ]```
becomes
` [ 'Selma',   'Lasem', 'Salem', 20, 'Elmas' ] `

ckandola commented:
@laurenclarke I've removed the duplicates now and just have the re-ordering problem left. Right now, all I can think of is using `split(' ')` on each name, i.e. `'FOREST PARK'.split(' ') ==> ['FOREST', 'PARK']` and checking if every other matched name includes everything in the array. But that sounds tedious. Is there a better way?

ckandola commented:
@prust I added Lauren's extensions, but it's just so messy-looking. 
I used uppercase names with no spaces and the method you suggested to get rid of duplicate names (with or without spaces), and then used split() and sort() on regular uppercase names and compared them to get rearranged words. Is there a better approach?

ckandola commented:
@laurenclarke It's still pretty messy-looking. What else should I do?

laurenclarke commented:
@ckandola 
I like how you are using git now, many small commits is so much better than a few big ones. 

>checking if every other matched name includes everything in the array. But that sounds tedious. Is there a better way?

Yes, I'd suggest doing at the word-level what you are doing at the letter-level to find the anagrams in the first place. For each name, split on space, sort, join and compare. Only this time you reject if there is a match (because this indicates that the names are the same words in possibly different order). 

> It's still pretty messy-looking. What else should I do?

Yes it is. It's good you recognize that. I think them main problem here is you have a  I think you have a few opportunities to make functions here even if they are called only once, putting code into a called function can help separate concerns (because you have to think carefully about what state gets passed to each), and keeps the overall template of what is happening clean while the details are elsewhere.

At the top of your file, you should have a short bit of set up and then a series of function calls to incrementally build the list. 

Before you start that, you might make a short recipe as series of comments. Something like..

```
// Setup
// Read source file
// Hash
// Find Anagrams
// Clean up Dupes
// Clean up word-for-word dupes (may want to do this before previous?)
// Write output
```
(not sure that's everything)
Then, under each comment you should have one or two function calls

Then below all that, you have your function definitions.

ckandola commented:
@laurenclarke 
>  For each name, split on space, sort, join and compare. Only this time you reject if there is a match (because this indicates that the names are the same words in possibly different order).
 
I never thought to sort after splitting! That helped a lot, and it allowed me to reuse a function instead of making a new one.

I tried separating the main parts into functions, so now there's a verifyAnagrams() function that takes the list and removes duplicates in steps. The main messy part now is in removeDuplicates, and it's just because I wanted to avoid using splice() on huge arrays. Is it looking okay now?

ckandola commented:
Okay, I fixed some of the rearranged words. Now, the results won't include:
'Bella Vista' and 'Vistabella'
'La Palme' and 'Palmela'
'Ia Pa' from ['Apia', 'Paia']
'Na Tan' and 'Tanna'
'Na Wa' and 'Wana'
'Monte Real' from ['Montelera', 'Realmonte']
'Park View' and 'Viewpark'
'Le Mans' from ['Mansle', 'Sleman', 'Meslan' ]
'Bonne Terre' and 'Terrebonne'
 'Crest Hill', and 'Hillcrest'
'el hed' and 'Hedel'
'Fort Pierre' and 'Pierrefort'
